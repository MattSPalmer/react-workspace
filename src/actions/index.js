export const splitItem = (id, newParentId, newItem, direction): Action => ({
  type: 'SPLIT_LAYOUT_ITEM',
  id,
  newParentId,
  newItem,
  direction: direction || 'perpendicular',
})

export const addItem = (id, item): Action => ({
  type: 'SPLIT_LAYOUT_ITEM',
  id,
  item,
})
