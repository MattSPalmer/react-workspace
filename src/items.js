/* eslint-disable react/require-render-return, class-methods-use-this */
import React from 'react'
import invariant from 'invariant'

export class Row extends React.Component {
  static createConfig = (elem) => ({
    type: 'row',
    id: elem.props.id,
    relativeScale: elem.props.relativeScale,
    innerProps: elem.props.innerProps || {},
    content: React.Children.map(elem.props.children, c => {
      if (!c.type.createConfig) {
        return null
      }
      return c.type.createConfig(c)
    })
  });

  render() {
    invariant(
      false,
      '<Row> elements are for Layout configuration only and should not be rendered'
    )
  }
}

export class Column extends React.Component {
  static createConfig = (elem) => ({
    type: 'column',
    id: elem.props.id,
    relativeScale: elem.props.relativeScale,
    innerProps: elem.props.innerProps || {},
    content: React.Children.map(elem.props.children, c => {
      if (!c.type.createConfig) {
        return null
      }
      return c.type.createConfig(c)
    })
  });

  render() {
    invariant(
      false,
      '<Column> elements are for Layout configuration only and should not be rendered'
    )
  }
}

export class Item extends React.Component {
  static createConfig = (elem) => ({
    type: 'item',
    id: elem.props.id,
    relativeScale: elem.props.relativeScale,
    component: elem.props.component,
    title: elem.props.title,
    innerProps: elem.props.innerProps || {},
  });

  render() {
    invariant(
      false,
      '<Item> elements are for Layout configuration only and should not be rendered'
    )
  }
}

