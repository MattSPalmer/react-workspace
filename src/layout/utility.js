import _ from 'lodash'

const recursiveTransform = transform => {
  function recursing(c) {
    if (!c.content) return transform(c)
    return transform({...c, content: c.content.map(recursing)})
  }
  return recursing
}

export function findOrientations(tree) {
  const {content} = tree
  if (!content) return tree

  const orientation = ({type}) => {
    switch (type) {
    case undefined:
    case 'row':
      return 'w'
    case 'column':
      return 'h'
    }
  }

  const mapper = c => findOrientations({
    ...c, orientation: orientation(tree)
  })
  return {...tree, content: content.map(mapper)}
}

function getDimensions(node) {
  const {parent, orientation, relativeScale} = node
  const {content, dim} = parent
  const parentDim = dim[orientation]

  let dims, noDimCount
  if (content) {
    dims = _(content).map('relativeScale').sum() * parentDim
    noDimCount = _(content).map('relativeScale').filter(_.isUndefined).value().length

    if (!dims || (dims > 1 || (noDimCount > 0 && dims === 1))) {
      dims = 0
      noDimCount = content.length
    }
  }

  let theDim = relativeScale ? relativeScale * parentDim : (parentDim - dims) / noDimCount
  switch (orientation) {
  case undefined:
  case 'w':
    return {...dim, w: theDim}
  case 'h':
    return {...dim, h: theDim}
  default:
    throw TypeError
  }
}

export function findDimensions(tree) {
  const {content} = tree
  if (!content) return tree

  const mapper = c => findDimensions({
    ...c, dim: getDimensions({...c, parent: tree})
  })
  return {...tree, content: content.map(mapper)}
}

export function findAspectRatios(tree) {
  const transform = c => ({
    ...c, dim: {...c.dim, ratio: c.dim.w / c.dim.h},
  })
  return recursiveTransform(transform)(tree)
}

export function transformAddIndex(tree) {
  const transform = (c, index = 0) => ({...c, index})
  return recursiveTransform(transform)(tree)
}

export function transformAddParent(tree) {
  const {content} = tree
  if (!content) return tree

  const mapper = c => transformAddParent({...c, parent: _.omit(tree, ['content'])})
  return {...tree, content: content.map(mapper)}
}

export function transformAddSiblings(tree) {
  const {content} = tree
  if (!content) return tree

  return {
    ...tree,
    content: content.map(
      c => ({...transformAddSiblings(c), siblings: content.length - 1})
    )
  }
}

function applyTransforms(tree, ...transforms) {
  for (let transform of transforms) {
    tree = transform(tree)
  }
  return tree
}

export function walkConfig(tree) {
  return applyTransforms(tree,
    findOrientations,
    findDimensions,
    findAspectRatios,
    transformAddIndex,
    transformAddSiblings,
    transformAddParent,
  )
}
