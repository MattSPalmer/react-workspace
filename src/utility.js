/* @flow */
/* eslint-disable no-console */
import _ from 'lodash'

type Props = { [key: string]: any }
type Dim = { w: number, h: number }

type Tree = {
  content: Array<Tree>,
  parent: ?Tree,
  relativeScale: ?number,
  innerProps: ?Props,
  dim: Dim,
  type: string,
}

const recursiveTransform = transform => {
  function recursing(c, index) {
    if (!c.content) return transform(c, index)
    return transform({...c, content: c.content.map(recursing)}, index)
  }
  return recursing
}

const orientationFromType = type => {
  switch (type) {
  case undefined:
  case 'row':
    return 'w'
  case 'column':
    return 'h'
  default:
    throw TypeError
  }
}

function getDimensions(node: Tree) {
  if (!node.parent) {
    return {w: 1, h: 1}
  }
  const {parent, relativeScale} = node
  const {content, dim, type} = parent
  const orientation = orientationFromType(type)
  const parentDim = dim[orientation]

  let dims, noDimCount
  if (content) {
    dims = _(content).map('relativeScale').sum() * parentDim
    noDimCount = _(content).map('relativeScale').filter(_.isUndefined).value().length

    if (!dims || (dims > 1 || (noDimCount > 0 && dims === 1))) {
      dims = 0
      noDimCount = content.length
    }
  }

  let theDim = relativeScale ? relativeScale * parentDim : (parentDim - dims) / noDimCount

  switch (orientation) {
  case undefined:
  case 'w':
    return {...dim, w: theDim}
  case 'h':
    return {...dim, h: theDim}
  default:
    throw TypeError
  }
}

export function findDimensions(tree: Tree) {
  const {content} = tree
  if (!content) return tree

  const mapper = c => findDimensions({
    ...c, dim: getDimensions({...c, parent: tree})
  })
  return {...tree, content: content.map(mapper)}
}

export function findAspectRatios(tree: Tree) {
  const transform = c => ({
    ...c, dim: {...c.dim, ratio: c.dim.w / c.dim.h},
  })
  return recursiveTransform(transform)(tree)
}

export function transformAddIndex(tree: Tree) {
  const transform = (c, index = 0) => ({...c, index})
  return recursiveTransform(transform)(tree)
}

export function transformAddParent(tree: Tree) {
  const {content} = tree
  if (!content) return tree

  const mapper = c => transformAddParent({...c, parent: _.omit(tree, ['content'])})
  return {...tree, content: content.map(mapper)}
}

export function transformAddSiblings(tree: Tree) {
  const {content} = tree
  if (!content) return tree

  return {
    ...tree,
    content: content.map(
      c => ({...transformAddSiblings(c), siblings: content.length - 1})
    )
  }
}

function applyTransforms(tree, ...transforms) {
  for (let transform of transforms) {
    tree = transform(tree)
  }
  return tree
}

function logFunc(tree: Tree) {
  const scale = tree.relativeScale || 1
  const parentScale = (tree.parent && tree.parent.relativeScale) || 1
  console.log(`Relative dimension is ${scale || 1}`)
  console.log(`Computed dimension is ${(scale || 1) * parentScale}`)
  if (tree.orientation) {
    console.log(`Orientation is ${tree.orientation}`)
  }
  console.log(`Dimensions are ${tree.dim.w}W x ${tree.dim.h}H`)
}

// eslint-disable-next-line no-unused-vars
function logWalk(tree: Tree) {
  console.groupCollapsed(tree.type || 'top')
  if (tree.parent)
    logFunc(tree)
  let newTree
  if (tree.content) {
    newTree = {...tree, content: tree.content.map(logWalk)}
  } else {
    newTree = tree
  }
  console.groupEnd()
  return newTree
}

export function walkConfig(tree: Tree) {
  return applyTransforms(tree,
    findOrientations,
    findDimensions,
    findAspectRatios,
    transformAddIndex,
    transformAddSiblings,
    transformAddParent,
  )
}
