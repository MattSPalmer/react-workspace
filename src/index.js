import React from 'react'
import {findDOMNode} from 'react-dom'
import {get, omit, debounce} from 'lodash'
import {LayoutParent, LayoutItem, NotFoundInRegister} from './renderedItems'
import {walkConfig} from './utility'

const Layout = React.createClass({
  getDefaultProps() {
    return {
      notFound: NotFoundInRegister,
      itemWrapper: 'div'
    }
  },
  getInitialState() {
    return {
      parentDim: { w: 0, h: 0 }
    }
  },
  renderLayoutChild(child) {
    const {register, itemWrapper, notFound} = this.props
    const key = child.id
    const {w, h} = this.state.parentDim
    const {dim} = child
    const [width, height] = [dim.w * w, dim.h * h].map(Math.floor)
    const reshapedProps = {
      ...child.innerProps,
      ...omit(child, 'innerProps'),
    }
    let commonProps = {key, width, height}
    if (child.type === 'item') {
      commonProps = {
        ...commonProps,
        componentClass: get(register, child.component, notFound),
        wrapper: get(register, itemWrapper, itemWrapper)
      }
      return <LayoutItem {...reshapedProps} {...commonProps} />
    }
    return (
      <LayoutParent {...child} {...commonProps}>
        {child.content && child.content.map(this.renderLayoutChild)}
      </LayoutParent>
    )
  },
  renderLayout() {
    const config = this.walkConfig()
    if (!config.content) return <div style={{display: 'flex'}}></div>
    return (
      <div style={{display: 'flex'}}>
        {config.content.map(this.renderLayoutChild)}
      </div>
    )
  },
  walkConfig() {
    const {config} = this.state
    return walkConfig(config)
  },
  createConfig(props = this.props) {
    const content = React.Children.map(props.children, c => {
      if (c.type.createConfig)
        return c.type.createConfig(c)
    })
    return {content, dim: {w: 1, h: 1}}
  },
  updateDimensions() {
    const elem = findDOMNode(this)
    const {offsetWidth, offsetHeight} = elem
    this.setState({
      parentDim: {w: offsetWidth, h: offsetHeight}
    })
  },
  componentWillMount() {
    this.setState({config: this.props.config || this.createConfig()})
  },
  componentDidMount() {
    const handleResize = debounce(this.updateDimensions, 100)
    window.addEventListener('resize', handleResize)
    this.updateDimensions()
  },
  componentWillReceiveProps(nextProps) {
    this.setState({
      config: nextProps.config || this.createConfig(nextProps)
    })
  },
  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize)
  },
  render() {
    return (
      <div>
        <div style={{width: '100%', height: '100%'}}>
          {this.renderLayout()}
        </div>
      </div>
    )
  }
})

export default Layout
