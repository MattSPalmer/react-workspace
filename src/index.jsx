import React from 'react'
import {findDOMNode} from 'react-dom'
import {get, omit, debounce} from 'lodash'
import {LayoutParent, LayoutItem, NotFoundInRegister} from './renderedItems'
import {walkConfig} from './utility'

class Layout extends React.Component {
  static defaultProps = {
    notFound: NotFoundInRegister,
    itemWrapper: 'div'
  };

  state = {
    parentDim: {w: 0, h: 0}
  };

  componentWillMount() {
    this.setState({config: this.props.config || this.createConfig()})
  }

  componentDidMount() {
    const handleResize = debounce(this.updateDimensions, 100)
    window.addEventListener('resize', handleResize)
    this.updateDimensions()
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      config: nextProps.config || this.createConfig(nextProps)
    })
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize)
  }

  updateDimensions = () => {
    const elem = findDOMNode(this)
    const {offsetWidth, offsetHeight} = elem
    this.setState({
      parentDim: {w: offsetWidth, h: offsetHeight}
    })
  };

  createConfig = (props = this.props) => {
    const content = React.Children.map(props.children, c => {
      if (!c.type.createConfig) {
        return null
      }
      return c.type.createConfig(c)
    })
    return {content, dim: {w: 1, h: 1}}
  };

  walkConfig = () => {
    const {config} = this.state
    return walkConfig(config)
  };

  renderLayoutChild = (child) => {
    const {register, itemWrapper, notFound} = this.props
    const key = child.id
    const {w, h} = this.state.parentDim
    const {dim} = child
    const [width, height] = [dim.w * w, dim.h * h].map(Math.floor)
    const reshapedProps = {
      ...child.innerProps,
      ...omit(child, 'innerProps'),
    }
    let commonProps = {key, width, height}
    if (child.type === 'item') {
      commonProps = {
        ...commonProps,
        componentClass: get(register, child.component, notFound),
        wrapper: get(register, itemWrapper, itemWrapper)
      }
      return <LayoutItem {...reshapedProps} {...commonProps} />
    }
    return (
      <LayoutParent {...child} {...commonProps}>
        {child.content && child.content.map(this.renderLayoutChild)}
      </LayoutParent>
    )
  };

  renderLayout = () => {
    const config = this.walkConfig()
    if (!config.content) return <div style={{display: 'flex'}} />
    return (
      <div style={{display: 'flex'}}>
        {config.content.map(this.renderLayoutChild)}
      </div>
    )
  };

  render() {
    return (
      <div>
        <div style={{width: '100%', height: '100%'}}>
          {this.renderLayout()}
        </div>
      </div>
    )
  }
}

const componentLike = React.PropTypes.oneOfType([
  React.PropTypes.string,
  React.PropTypes.func,
  React.PropTypes.instanceOf(React.Component),
])

Layout.propTypes = {
  register: React.PropTypes.object,
  config: React.PropTypes.object,
  itemWrapper: componentLike,
  notFound: componentLike,
}

export default Layout
