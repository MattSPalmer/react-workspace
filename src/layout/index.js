import React from 'react'
import Pane from './pane'
import {walkConfig} from './utility'

const Layout = React.createClass({
  getInitialState() {
    return {}
  },
  renderItem() {

  },
  renderItems() {
    console.log(
      JSON.stringify(this.walkConfig(), null, 2)
    )
  },
  walkConfig() {
    const {config} = this.state
    return walkConfig(config)
  },
  createConfig() {
    const content = React.Children.map(this.props.children, c => {
      if (c.type.createConfig)
        return c.type.createConfig(c)
    })
    return {content}
  },
  componentWillMount() {
    this.setState({config: this.createConfig()})
  },
  render() {
    return (
      <div>
        <div style={{width: '100%', height: '100%'}}>
          {this.renderItems()}
        </div>
      </div>
    )
  }
})

export default Layout
