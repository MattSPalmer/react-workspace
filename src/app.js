import React from 'react'

import Layout from './layout'
import {Row, Column, Item} from './layout/items'
import Pane from './layout/pane'

const Box = ({color = 'black', meta}) => (
  <div style={{padding: 10, width: '100%', height: '100%'}}>
    <div>
      <h2 style={{margin: 0}}>
        {meta.width} x {meta.height}
      </h2>
      <pre style={{margin: 0}}>
        {JSON.stringify(meta, null, 2)}
      </pre>
    </div>
  </div>
)

const Sizing = React.createClass({
  coloredDim(dim, threshold = 1000) {
    const style = {
      color: dim > threshold ? '#F99' : 'unset'
    }
    return <span style={style}>{dim}</span>
  },
  render() {
    const {meta} = this.props
    const wrapStyle = {
      width: '100%', height: '100%',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    }
    const headerStyle = {
      color: '#CCC', fontSize: meta.width / 7,
      margin: 0, display: 'inline-block',
    }
    return (
      <div style={wrapStyle}>
        <h1 style={headerStyle}>
          {this.coloredDim(meta.width)} x {this.coloredDim(meta.height)}
        </h1>
      </div>
    )
  }
})

const Timer = React.createClass({
  getInitialState() {
    return {time: 0}
  },
  componentDidMount() {
    const fn = () => this.setState({time: this.state.time + 1})
    this.setState({
      timerId: window.setInterval(fn, 1000)
    })
  },
  componentWillUnmount() {
    window.clearInterval(this.state.timerId)
  },
  handleClick(e) {
    this.setState({time: 0})
  },
  formattedTime() {
    const {time} = this.state
    const zeroPad = (input, length = 2) => {
      input = "" + input
      let paddingSize = Math.max(0, length - input.length)
      return new Array(paddingSize > 0 ? paddingSize + 1 : 0).join("0") + input;
    }
    const [min, sec] = [Math.floor(time / 60), time % 60]
    return `${zeroPad(min)}:${zeroPad(sec)}`
  },
  render() {
    const {meta} = this.props
    const wrapStyle = {
      width: '100%', height: '100%',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    }
    const headerStyle = {
      color: this.props.color || 'black', fontSize: meta.width / 3,
      margin: 0, display: 'inline-block',
    }
    return (
      <div style={wrapStyle}>
        <h2 style={headerStyle} onClick={this.handleClick}>{this.formattedTime()}</h2>
      </div>
    )
  }
})

const thing = (
  <Column>
    <Row>
      <Item component='size' />
      <Item component='size' />
    </Row>
    <Item component='size' />
    <Item component='size' />
  </Column>
)

export default React.createClass({
  render() {
    const register = {
      size: Sizing, 
      time: Timer,
      debug: Box
    }
    return (
      <div style={style.app}>
        <Layout register={register} itemWrapper={Pane}>
          {thing}
        </Layout>
      </div>
    )
  }
})

const style = {
  app: {
    width: '100%',
    position: 'relative',
    float: 'left',
    fontFamily: 'Open Sans',
  },
}

