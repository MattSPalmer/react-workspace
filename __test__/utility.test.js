import * as u from '../src/utility'

describe('tree manipulation', () => {
  const inputWithChildren = {
    id: 'root',
    dim: {w: 1, h: 1},
    content: [
      {
        id: 'row1',
        type: 'row',
        content: [
          {id: 'row1child1'},
          {id: 'row1child2'},
          {id: 'row1child3'},
        ],
      },
      {
        id: 'column1',
        type: 'column',
        content: [
          {id: 'column1child1'},
          {id: 'column1child2'},
          {id: 'column1child3'},
        ],
      },
    ],
  }
  describe('findDimensions', () => {
    it('finds the dimensions of a basic tree', () => {
      const output = u.findDimensions(inputWithChildren)
      expect(output).toMatchSnapshot()
    })

    it('correctly calculates width or height based on parent', () => {
      const output = u.findDimensions(inputWithChildren)
      expect(output).toMatchSnapshot()
    })
  })

  it('adds the index', () => {
    const output = u.transformAddIndex(inputWithChildren)
    expect(output).toMatchSnapshot()
  })

  it('adds aspect ratios', () => {
    let output = u.findDimensions(inputWithChildren)
    output = u.findAspectRatios(output)

    expect(output).toMatchSnapshot()
  })
})
