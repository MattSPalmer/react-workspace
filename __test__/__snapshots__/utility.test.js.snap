exports[`tree manipulation adds aspect ratios 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "dim": Object {
            "h": 1,
            "ratio": 0.16666666666666666,
            "w": 0.16666666666666666
          },
          "id": "row1child1"
        },
        Object {
          "dim": Object {
            "h": 1,
            "ratio": 0.16666666666666666,
            "w": 0.16666666666666666
          },
          "id": "row1child2"
        },
        Object {
          "dim": Object {
            "h": 1,
            "ratio": 0.16666666666666666,
            "w": 0.16666666666666666
          },
          "id": "row1child3"
        }
      ],
      "dim": Object {
        "h": 1,
        "ratio": 0.5,
        "w": 0.5
      },
      "id": "row1",
      "type": "row"
    },
    Object {
      "content": Array [
        Object {
          "dim": Object {
            "h": 0.3333333333333333,
            "ratio": 1.5,
            "w": 0.5
          },
          "id": "column1child1"
        },
        Object {
          "dim": Object {
            "h": 0.3333333333333333,
            "ratio": 1.5,
            "w": 0.5
          },
          "id": "column1child2"
        },
        Object {
          "dim": Object {
            "h": 0.3333333333333333,
            "ratio": 1.5,
            "w": 0.5
          },
          "id": "column1child3"
        }
      ],
      "dim": Object {
        "h": 1,
        "ratio": 0.5,
        "w": 0.5
      },
      "id": "column1",
      "type": "column"
    }
  ],
  "dim": Object {
    "h": 1,
    "ratio": 1,
    "w": 1
  },
  "id": "root"
}
`;

exports[`tree manipulation adds the index 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "id": "row1child1",
          "index": 0
        },
        Object {
          "id": "row1child2",
          "index": 1
        },
        Object {
          "id": "row1child3",
          "index": 2
        }
      ],
      "id": "row1",
      "index": 0,
      "type": "row"
    },
    Object {
      "content": Array [
        Object {
          "id": "column1child1",
          "index": 0
        },
        Object {
          "id": "column1child2",
          "index": 1
        },
        Object {
          "id": "column1child3",
          "index": 2
        }
      ],
      "id": "column1",
      "index": 1,
      "type": "column"
    }
  ],
  "dim": Object {
    "h": 1,
    "w": 1
  },
  "id": "root",
  "index": 0
}
`;

exports[`tree manipulation findDimensions correctly calculates width or height based on parent 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "dim": Object {
            "h": 1,
            "w": 0.16666666666666666
          },
          "id": "row1child1"
        },
        Object {
          "dim": Object {
            "h": 1,
            "w": 0.16666666666666666
          },
          "id": "row1child2"
        },
        Object {
          "dim": Object {
            "h": 1,
            "w": 0.16666666666666666
          },
          "id": "row1child3"
        }
      ],
      "dim": Object {
        "h": 1,
        "w": 0.5
      },
      "id": "row1",
      "type": "row"
    },
    Object {
      "content": Array [
        Object {
          "dim": Object {
            "h": 0.3333333333333333,
            "w": 0.5
          },
          "id": "column1child1"
        },
        Object {
          "dim": Object {
            "h": 0.3333333333333333,
            "w": 0.5
          },
          "id": "column1child2"
        },
        Object {
          "dim": Object {
            "h": 0.3333333333333333,
            "w": 0.5
          },
          "id": "column1child3"
        }
      ],
      "dim": Object {
        "h": 1,
        "w": 0.5
      },
      "id": "column1",
      "type": "column"
    }
  ],
  "dim": Object {
    "h": 1,
    "w": 1
  },
  "id": "root"
}
`;

exports[`tree manipulation findDimensions finds the dimensions of a basic tree 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "dim": Object {
            "h": 1,
            "w": 0.16666666666666666
          },
          "id": "row1child1"
        },
        Object {
          "dim": Object {
            "h": 1,
            "w": 0.16666666666666666
          },
          "id": "row1child2"
        },
        Object {
          "dim": Object {
            "h": 1,
            "w": 0.16666666666666666
          },
          "id": "row1child3"
        }
      ],
      "dim": Object {
        "h": 1,
        "w": 0.5
      },
      "id": "row1",
      "type": "row"
    },
    Object {
      "content": Array [
        Object {
          "dim": Object {
            "h": 0.3333333333333333,
            "w": 0.5
          },
          "id": "column1child1"
        },
        Object {
          "dim": Object {
            "h": 0.3333333333333333,
            "w": 0.5
          },
          "id": "column1child2"
        },
        Object {
          "dim": Object {
            "h": 0.3333333333333333,
            "w": 0.5
          },
          "id": "column1child3"
        }
      ],
      "dim": Object {
        "h": 1,
        "w": 0.5
      },
      "id": "column1",
      "type": "column"
    }
  ],
  "dim": Object {
    "h": 1,
    "w": 1
  },
  "id": "root"
}
`;
